projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -4,222 +4,263 @@ import apiFetch from '@wordpress/api-fetch';
+			</div>
+		</>
+	);
+}
 
-	// No content, only title and categories
-	if ( categoryNames.length ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and comma-seperated category names. */
-			__( "This is a post titled '%1$s', published in categories '%2$s':\n", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			categoryNames
-		);
-	}
+export default function Edit( { attributes, setAttributes, clientId } ) {
+	const [ loadingCompletion, setLoadingCompletion ] = useState( false );
+	const [ loadingCategories, setLoadingCategories ] = useState( false );
+	const [ showAnimation, setShowAnimation ] = useState( false );
+	const [ needsMoreCharacters, setNeedsMoreCharacters ] = useState( 0 );
+	const [ showRetry, setShowRetry ] = useState( false );
+	const [ completionFinished, setCompletionFinished ] = useState( !! attributes.content );
+	const [ errorMessage, setErrorMessage ] = useState( false );
 
-	// Title and content
-	if ( content ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and the last 240 characters of content. */
-			__( "This is a post titled '%1$s':\n\n…%2$s", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			content
-		);
-	}
+	// Let's grab post data so that we can do something smart.
+	const currentPostTitle = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'title' ),
+		[]
+	);
 
-	return sprintf(
-		/** translators: This will be a prompt to OpenAI to generate a post based on the post title */
-		__( 'Write content of a post titled "%s"', 'jetpack' ),
-		title
+	let loading = false;
+	const categories = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ),
+		[]
 	);
-}
 
-/**
- * Gather data available in Gutenberg and prepare the best prompt we can come up with.
- *
- * @param {Function} select - function returning Gutenberg data store.
- * @returns {string} - prompt ready to pipe to OpenAI.
- */
-function preparePromptBasedOnEditorState( select ) {
-	const editorContent = formatEditorCanvasContentForCompletion(
-		select( 'core/block-editor' )
-	).slice( -240 );
+	const categoryObjects = useSelect( select => {
+		return categories
+			.map( categoryId => {
+				const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
 
-	// Let's grab post data so that we can do something smart.
-	const currentPost = select( 'core/editor' ).getCurrentPost();
-	if ( ! currentPost ) {
-		return '';
-	}
+				if ( ! category ) {
+					// Data is not yet loaded
+					loading = true;
+					return;
+				}
 
-	// If there is no title, there is not much we can do.
-	if ( ! currentPost.title ) {
-		return createPrompt( '', editorContent, '' );
-	}
+				return category;
+			} )
+			.filter( Boolean ); // Remove undefined values
+	} );
```

Comment:
I haven't tested anything, but here you should have `categories` as deps to `useSelect` and it's better to fetch all terms and `map+filter` outside the `useSelect` because these functions return a new `array` instance.

Reply:
>  it's better to fetch all terms 

I know, but we haven't been able to fetch all terms... Discussion: p1674231456087539/1674220033.464909-slack-C02A41GCS

Reply:
> t here you should have categories as deps to useSelect

You're right

> because these functions return a new array instance.

That doesn't really matter as `categoryObjects` is not a dependency for anything...

Reply:
Done in 58c3d1508dfd0efb90d0103e27ba4f44bdf853f4